### PR TITLE
Allow mapping parameters to joined fields

### LINKED ISSUE #11409: Dashboard filters can not filter joined tables
**Describe the bug**
Fields in joined tables in custom questions are not available as dashboard filters.

**To Reproduce**
1. Create custom question for Sample database
2. From order, join products on productid and review on productid.
3. Summarize on count of product name
4. Save and add to dashboard
5. Create dashboard filter and filter on review rating (not possible as it doesn’t show up in the drop down)

**Expected behavior**
Fields in joined tables should be available in dashboard filters.

**Screenshots**
Filter on reviews

**Information about your Metabase Installation:**
0.33.6
`
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "Java(TM) SE Runtime Environment",
    "java.runtime.version": "1.8.0_144-b01",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_144",
    "java.vm.name": "Java HotSpot(TM) 64-Bit Server VM",
    "java.vm.version": "25.144-b01",
    "os.name": "Windows 10",
    "os.version": "10.0",
    "user.language": "en",
    "user.timezone": "Europe/Berlin"
  },
  "metabase-info": {
    "databases": [
      "mysql"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "run-mode": "prod",
    "version": {
      "tag": "v0.33.6",
      "date": "2019-11-19",
      "branch": "release-0.33.x",
      "hash": "be1e0e1"
    },
    "settings": {
      "report-timezone": "Europe/Berlin"
    }
  }
}
`
**Severity**
Annoying. The workaround is to create a native SQL query which supports these queries. 

**Additional context**
Also see: https://discourse.metabase.com/t/dashboard-filters-cant-filter-on-joined-fields-for-custom-questions/8175

### PR DESCRIPTION
**Pull request by tlrobinson**
* This changes the parameter mapping code to look at the query's `dimensionOptions()` for available dimensions, which includes joined fields.
* It adds `dimensionOptions()` to `NativeQuery` (`StructuredQuery` already had it) and adds a `TemplateTagDimension`
* It also introduces a first class `Variable` object, of which there's only currently the `TemplateTagVariable` type returned by `NativeQuery`'s `variables()` (but you could imagine `StructuredQuery` having variables that can be interpolated in MBQL as well)

Resolves #11409

(I plan on adding some tests)

### CODE REVIEW

File: frontend/src/metabase/meta/Dashboard.js [hunk 1]
```diff
@@ -312,48 +197,40 @@ function tagFilterForParameter(parameter: Parameter): TemplateTagFilter {
...
+      .map(section =>
+        section.items.map(({ dimension }) => ({
+          name: dimension.displayName(),
+          target: ["dimension", dimension.mbql()],
+          icon: dimension.icon(),
+          sectionName: section.name,
+          isFk: section.icon !== "table2", // HACK: super fragile
+        })),
+      )
+      .flat(),
```
Comment (paulrosenzweig) [inline frontend/src/metabase/meta/Dashboard.js]: `.flatMap`?

File: frontend/src/metabase/meta/Dashboard.js [hunk 2]
```diff
@@ -312,48 +197,40 @@ function tagFilterForParameter(parameter: Parameter): TemplateTagFilter {
...
+    ...query
+      .dimensionOptions(dimensionFilterForParameter(parameter))
+      .sections()
+      .map(section =>
+        section.items.map(({ dimension }) => ({
+          name: dimension.displayName(),
+          target: ["dimension", dimension.mbql()],
+          icon: dimension.icon(),
+          sectionName: section.name,
+          isFk: section.icon !== "table2", // HACK: super fragile
```
Comment (paulrosenzweig) [inline frontend/src/metabase/meta/Dashboard.js]: Why can't this be set from `dimension.field().isFK()?`?

**Review by paulrosenzweig — APPROVED** [review]
Looks great! Much nicer structure than before where the dimension options logic was partially duplicated in dashboards.

**Review by kdoh — APPROVED** [review]
Looks like things are working as expected 👍

**Comment by jakedeg** [comment]
Really looking forward to this. Any idea which release it might be merged into?

**Comment by flamber** [comment]
@jakedeg This has been merged and part of Metabase since 0.34.2

**Comment by jakedeg** [comment]
Ah! Fantastic. We just upgraded from 0.34.0 to 0.34.3. Gonna check it out now. Thanks for letting me know. I was having trouble finding it in release notes or understanding which version it was slated for.

**Comment by jakedeg** [comment]
omg. heaven. this is *so* helpful for the dashboard we're making. thank you!

**Comment by jakedeg** [comment]
One follow-up question on this - I'm now using these joined fields for Category filters on a dashboard, but I'm finding the inputs for those filters are remaining text input and i'm not seeing the Search Box or drop-down as described [here](https://www.metabase.com/docs/latest/users-guide/08-dashboard-filters.html). If I type in the exactly right value, the filter works, but a drop-down or searchable list or auto-complete would be far easier to use.

Should we expect select inputs to work with these sorts of filters and I'm just doing something wrong? Or is this a limitation of filters based on joined fields?

**Comment by flamber** [comment]
@jakedeg 
It depends on what you've set in Data Model and if a filter is connected to multiple different questions.
Please use the forum for questions and troubleshooting: https://discourse.metabase.com/

**Comment by flamber** [comment]
@jakedeg Seems like someone reported this #12250

**Pull request merged**